hare/algorithm.go

```diff
@@ -330,17 +351,36 @@ PreRound:
 
 	// start first iteration
 	proc.onRoundBegin(ctx)
-	endOfRound = proc.clock.AwaitEndOfRound(proc.k)
+	endOfRound = proc.clock.AwaitEndOfRound(proc.getK())
+
+	defer func() {
+		if !proc.completed {
+			proc.report(notCompleted)
+		}
+	}()
 
 	for {
 		select {
 		case msg := <-proc.inbox: // msg event
+
 			proc.handleMessage(ctx, msg)
-			if proc.terminating {
+
+			if proc.terminating || (proc.completed && proc.certified) {
+				// TODO: possible error close closed channel
```

Comment:
I think this is not possible, because there's a return after every close and the event loop only runs once. If you think it's possible - please explain (we can fix it by checking for a closed channel in `proc.Close()` before closing, but I don't think it's needed today). Otherwise, let's remove the `TODO`.